sdk/metric/aggregator/histogram/histogram.go

```diff
@@ -117,30 +113,42 @@ func (c *Aggregator) Checkpoint(ctx context.Context, desc *metric.Descriptor) {
 	c.lock.Unlock()
 }
 
-func emptyState(boundaries []metric.Number) state {
+func emptyState(boundaries []float64) state {
 	return state{
-		bucketCounts: make([]metric.Number, len(boundaries)+1),
+		bucketCounts: make([]float64, len(boundaries)+1),
 	}
 }
 
 // Update adds the recorded measurement to the current data set.
 func (c *Aggregator) Update(_ context.Context, number metric.Number, desc *metric.Descriptor) error {
 	kind := desc.NumberKind()
+	asFloat := number.CoerceToFloat64(kind)
 
 	bucketID := len(c.boundaries)
 	for i, boundary := range c.boundaries {
-		if number.CompareNumber(kind, boundary) < 0 {
+		if asFloat < boundary {
 			bucketID = i
 			break
 		}
 	}
+	// Note: Binary-search was compared using the benchmarks. The following
+	// code is equivalent to the linear search above:
+	//
+	//     bucketID := sort.Search(len(c.boundaries), func(i int) bool {
+	//         return asFloat < c.boundaries[i]
+	//     })
+	//
+	// The binary search wins for very large boundary sets, but
+	// the linear search performs better up through arrays between
+	// 256 and 512 elements, which is a relatively large histogram, so we
+	// continue to prefer linear search.
```

Comment:
Should the optimization be added to use the Binary search when the length exceeds `256` or `512`?

Reply:
I thought about it. It's a pretty big array of values for a histogram, and probably at that size we should be talking about log-linear or linear buckets, which support better ways to search. I guess I think probably not, but can be convinced.

Reply:
I think you are correct this would be a rare edge case. It is an optimization we can add in the future if we discover it is needed.

:+1: